src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -894,101 +934,326 @@ private bool CheckPropertyValueKind(SyntaxNode node, BoundExpression expr, BindV
+                    var valid = effectiveRefKind != RefKind.None && argument.Type.IsByRefLikeType == false ?
+                                        CheckRefEscape(argument.Syntax, argument, escapeFrom, escapeTo, false, diagnostics) :
+                                        CheckValEscape(argument.Syntax, argument, escapeFrom, escapeTo, false, diagnostics);
+
+                    if (!valid)
                     {
-                        var errorCode = checkingReceiver ? ErrorCode.ERR_RefReturnCall2 : ErrorCode.ERR_RefReturnCall;
-                        var parameterIndex = argToParamsOpt.IsDefault ? argIndex : argToParamsOpt[argIndex];
-                        var parameterName = parameters[parameterIndex].Name;
+                        // no longer needed
+                        inParametersMatchedWithArgs?.Free();
+
+                        ErrorCode errorCode = GetStandardCallEscapeError(checkingReceiver);
+                        var paramIndex = argsToParamsOpt.IsDefault ? argIndex : argsToParamsOpt[argIndex];
+                        var parameterName = parameters[paramIndex].Name;
                         Error(diagnostics, errorCode, syntax, symbol, parameterName);
                         return false;
                     }
                 }
+            }
 
-                return true;
+            // handle omitted optional "in" parameters if there are any
+            ParameterSymbol unmatchedInParameter = TryGetUnmatchedInParameterAndFreeMatchedArgs(parameters, ref inParametersMatchedWithArgs);
+
+            // unmatched "in" parameter is same as a literal, its ref escape is scopeOfTheContainingExpression  (can't get any worse)
+            //                                                its val escape is ExternalScope                   (does not affect overal result)
+            if (unmatchedInParameter != null)
+            {
+                Error(diagnostics, GetStandardCallEscapeError(checkingReceiver), syntax, symbol, unmatchedInParameter.Name);
+                return false;
             }
 
-            // check all "in" parameters 
-            for (var paramIndex = 0; paramIndex < parameters.Length; paramIndex++)
+            // check receiver if ref-like
+            if (receiverOpt?.Type?.IsByRefLikeType == true)
             {
-                var parameter = parameters[paramIndex];
+                return CheckValEscape(receiverOpt.Syntax, receiverOpt, escapeFrom, escapeTo, false, diagnostics);
+            }
 
-                if (parameter.RefKind != RefKind.RefReadOnly)
-                {
-                    continue;
-                }
+            return true;
+        }
+
+        /// <summary>
+        /// Validates whether the invocation is valid per no-mixing rules.
+        /// Returns `false` when it is not valid and produces diagnostics (possibly more than one recursively) that helps to figure the reason.
+        /// </summary>
+        private static bool CheckInvocationArgMixing(
```

Comment:
The spec doesn't explain arg mixing, as far as I can tell. Can you take a note to document? Thanks

Reply:
The spec has no-mixing under:
* In a method invocation, the following constraints apply:

---
In reply to: [135907546](https://github.com/dotnet/roslyn/pull/21334#discussion_r135907546) [](ancestors = 135907546)